src/cancel_order.rs

```diff
@@ -1,6 +1,198 @@
+                .with_static_gas(Gas(10))
+                .with_attached_deposit(1)
+                .remove_liquidity(
+                    order.lpt_id.clone(),
+                    U128(amount),
+                    U128(min_amount_x),
+                    U128(min_amount_y),
+                )
+                .then(
+                    ext_self::ext(current_account_id())
+                        .with_static_gas(Gas(5))
+                        .with_attached_deposit(NO_DEPOSIT)
+                        .remove_liquidity_callback(order_id, swap_fee, price_impact),
+                );
+        } else {
+            self.swap(order_id, swap_fee, price_impact);
+        }
+    }
+
+    #[private]
+    pub fn remove_liquidity_callback(&self, order_id: U128, swap_fee: U128, price_impact: U128) {
+        require!(is_promise_success(), "Some problem with remove liquidity");
+        self.swap(order_id, swap_fee, price_impact);
+    }
+
+    fn swap(&self, order_id: U128, swap_fee: U128, price_impact: U128) {
+        let orders = self.orders.get(&signer_account_id()).unwrap_or_else(|| {
+            panic!("Orders for account: {} not found", signer_account_id());
+        });
+
+        let order = orders.get(&(order_id.0 as u64)).unwrap_or_else(|| {
+            panic!("Order with id: {} not found", order_id.0);
+        });
+
+        let buy_amount =
+            BigDecimal::from(order.amount) * order.leverage * order.sell_token_price.value
+                / order.buy_token_price.value;
+        let min_amount = buy_amount * self.get_price(order.buy_token.clone());
+        let actions: Vec<Action> = vec![Action::Swap(SwapAction {
+            pool_id: self.pool_id,
+            token_in: order.buy_token.clone(),
+            amount_in: Some(WRatio::from(buy_amount)),
+            token_out: order.sell_token.clone(),
+            min_amount_out: WRatio::from(min_amount),
+        })];
+        let action = TokenReceiverMessage::Execute {
+            force: true,
+            actions,
+        };
+
+        log!(
+            "action {}",
+            near_sdk::serde_json::to_string(&action).unwrap()
+        );
+
+        ext_token::ext(order.buy_token.clone())
+            .with_static_gas(Gas(3))
+            .with_attached_deposit(1)
+            .ft_transfer_call(
+                self.ref_finance_account.clone(),
+                WRatio::from(buy_amount),
+                Some("Deposit tokens".to_string()),
+                near_sdk::serde_json::to_string(&action).unwrap(),
+            )
+            .then(
+                ext_self::ext(current_account_id())
+                    .with_static_gas(Gas(20))
+                    .with_attached_deposit(NO_DEPOSIT)
+                    .cancel_swap_callback(order_id, swap_fee, price_impact),
+            );
+    }
+
+    #[private]
+    pub fn cancel_swap_callback(&mut self, order_id: U128, swap_fee: U128, price_impact: U128) {
```

Comment:
@TymofiiSavinov why it's called `cancel_swap_callback` if the handler functions `swap`?

Reply:
we have swap_callback in create order, but with not same functionality. I changed name to order_cancel_swap_callback